node/core/pvf/src/prepare/worker.rs

```diff
@@ -314,59 +307,61 @@ async fn recv_request(stream: &mut UnixStream) -> io::Result<(Vec<u8>, PathBuf,
 /// The entrypoint that the spawned prepare worker should start with. The `socket_path` specifies
 /// the path to the socket used to communicate with the host.
 pub fn worker_entrypoint(socket_path: &str) {
-	worker_event_loop("prepare", socket_path, |mut stream| async move {
+	worker_event_loop("prepare", socket_path, |stream| async move {
+		let mutex = Arc::new(Mutex::new((stream, false)));
```

Comment:
This looks really horrible. :see_no_evil: 

You should put the stream into its own future and then have a channel to communicate with it.

Reply:
From the cursory look there might be something I'm not seeing here, but can't we just simplify this code and get rid of this `Mutex` and the flag by:

1. Have only this task handle the communication on the socket.
2. Have it spawn two tasks: 1) doing the work, 2) monitoring the cpu time.
3. `select!` the task that finishes first.
4. Handle the result.

?

Reply:
Yeah. Basically what I also wanted to achieve with "its own future and then have a channel to communicate with it".

But yours sounds more clean :D

Reply:
@koute That sounds really neat, but I think we would need two `spawn_blocking` calls, as the cpu time loop and artifact preparation are both blocking. It would be better code clarity for sure, at the expense of another thread.

@bkchr I'm not sure what your solution would look like. Again we have the issue where the main thread (doing the preparation/execution) is blocked, so we can't run any other tasks on it. That's why this looks so ugly. :/ Let me know if I'm missing something.

Reply:
> @koute That sounds really neat, but I think we would need two `spawn_blocking` calls, as the cpu time loop and artifact preparation are both blocking. It would be better code clarity for sure, at the expense of another thread.

Should be fine or what is the problem you see?

Reply:
I don't have any problems with mutexing the socket:)

Reply:
Should be fine spawning the extra thread.

1. Things like PVF precompilation will absolutely dwarf the extra overhead anyway.
2. If we want to be frugal with the resources (with the number of launched threads) we should probably first change how we spawn the runtime, as in the default configuration that will spawn a lot more threads than the extra one. (:

Reply:
> Which threads do we spawn?

We do this:

```rust
thread::Builder::new().name("CPU time monitor".into()).spawn(move || {
	let rt = Runtime::new().unwrap();
```

This by default immediately spawns as many threads as you have hardware threads, e.g. 32 core/64 thread CPU will get 64 threads spawned.

Reply:
Does anything stop us from making this code sync and removing tokio runtime completely? There's nowhere to yield in case of a blocking call anyway, 1 threads sleeps-polls, another one performs blocking preparation

Reply:
> > Which threads do we spawn?
> 
> We do this:
> 
> ```rust
> thread::Builder::new().name("CPU time monitor".into()).spawn(move || {
> 	let rt = Runtime::new().unwrap();
> ```
> 
> `Runtime::new` by default immediately spawns as many threads as you have hardware threads, e.g. 32 core/64 thread CPU will get 64 threads spawned.

Ahh you meant here in this code :P And yeah that is a good point. I still think that @koute plan is the best way forward.

Reply:
> Does anything stop us from making this code sync and removing tokio runtime completely? There's nowhere to yield in case of a blocking call anyway, 1 threads sleeps-polls, another one performs blocking preparation

Theoretically yes, but it would be a bigger refactor and we would need multiple threads anyway.

Reply:
@koute Oof, TIL. But those two lines are being replaced by `spawn_blocking`. But yeah, if the extra thread is fine then the `select` version is easier to read, I will do that.

@slumber It can be sync! We just have to replace the async sockets with the ones from the std library.

Reply:
Went ahead with the async version with `select!`.